Optimize assert mechanism

joerchan opened the pull request:
__assert
--------
Add Kconfig option to disable the conditional expression in the assert
that failed. This would save code space, and file and line provides
better information than the conditional expression in case where
the same expression would be asserted upon.

For example __ASSERT_NO_MSG(buf) wouldn't make much sense in
configuration where CONFIG_ASSERT_NO_FILE_INFO was enabled.

Add option to disable the assertion message, this makes all __ASSERT
behave as __ASSERT_NO_MSG instead.

Add verbose option which would control if the assertion mechanism prints
any information at all. With this disabled they application will have to
use the stack-frame to locate the assertion location.

Add option to disable the assertion message, this makes all __ASSERT
behave as __ASSERT_NO_MSG instead.

Add verbose option which would control if the assertion mechanism prints
any information at all. With this disabled they application will have to
use the stack-frame to locate the assertion location.

Disable printing the line number in assertions when file name has been
disabled. Knowing the line number is not very useful when the name of
the file is unknown.

k_panic
----------
Use __ASSERT_LOC to print panic or oops location, this enables the panic
and oops handling to disable printing file and line information. The
cause of the exception can be backtraced using the stackframe instead,
which would give the user a way to have a minimal footprint panic
implementation.

Bluetooth and net_buf
-------------------------
Change assertion messaged printed from printing expression to printing
the line and file name. This provides more context as the same
expression could be asserted upon multiple times and would then not
provide enough clarity in the message.

Also using a formatted string would save code space as we can use the
same string for all messages instead of creating a unique one for each
condition.

Removed NET_BUF_ASSERT an using __ASSERT instead.

ROM comparison
-------------------
Tests results using `tests/bluetooth/shell`

2.1.0 tag
--------

Without asserts
**ROM report:** 235488 

With asserts configuration:
```
CONFIG_ASSERT=y
CONFIG_BT_ASSERT=y
CONFIG_BT_ASSERT_VERBOSE=y
CONFIG_NET_BUF_LOG=y
```
**ROM report:** 239156


This PR:
--------
Without asserts
**ROM report:** 234800

With assert configuration:
```
CONFIG_ASSERT=y
CONFIG_ASSERT_VERBOSE=y
CONFIG_ASSERT_NO_FILE_INFO=n
```
**ROM report:** 240591

With assert configuration:
```
CONFIG_ASSERT=y
CONFIG_ASSERT_VERBOSE=n
CONFIG_ASSERT_NO_FILE_INFO=y

```
**ROM report:** 216551

~~I can't really explain why enabling all the assert options decreases the
code usage right now. Anybody who are experts in analyzing elf files who can say?~~
With the options it is now possible to disable more string additions to the final elf file so comparing becomes less obvious. Some of the options can be used to reduce the final image size without enabling asserts as well. 

EDIT: Updated description and rom_report numbers after introducing more optimizations and options.

joerchan committed:
assert: Add option to disable condition from assertion message

Add Kconfig option to disable the conditional expression in the assert
that failed. This would save code space, and file and line provides
better information than the conditional expression in case where
the same expression would be asserted upon.

For example __ASSERT_NO_MSG(buf) wouldn't make much sense in
configuration where CONFIG_ASSERT_NO_FILE_INFO was enabled.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: assert: Change printed expression to printing line and file

Change assertion messaged printed from printing expression to printing
the line and file name. This provides more context as the same
expression could be asserted upon multiple times and would then not
provide enough clarity in the message.

Also using a formatted string would save code space as we can use the
same string for all messages instead of creating a unique one for each
condition.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
net: buf: Switch from net buf assert to system assert

Remove net bufs own assertion mechanism and use the system assert
instead. This changes the assertion messaged printed from printing
expression to printing the line and file name. This provides more
context as the same expression could be asserted upon multiple times and
would then not provide enough clarity in the message.

This removes the option to enable only net buf assertions.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

rakons commented:
> I can't really explain why enabling all the assert options decreases the
> code usage right now. Anybody who are experts in analyzing elf files who can say?

As assert makes the program to die, and not continuing, it may be due to some duplication of the assertion checks in the code - for example first the assertion checks if a == 0 and then in the code we have normal code flow that checks if a < 0 - this may be removed if the assertion is enabled. But also it may mean that there is something really broken somewhere in the code logic.

Just an idea - maybe somebody would come with another explanation.

joerchan commented:
> While this will save some space as your metrics show, @joerchan, if you are not going to print any info then why not just switch to `k_panic()`? That should resolve in the compiler injecting a controlled fault that can resolve to a PC that can then be used to track down the location in the source code that triggered the `k_panic()`. Could you try that and see if it works and yields better ROM footprint?
> CC @ioannisg @andrewboie

@carles: So what we would like to do is to have CONFIG_BT_ASSERT_VERBOSE similar for __ASSERT and NET_BUF then?

carlescufi commented:
> > While this will save some space as your metrics show, @joerchan, if you are not going to print any info then why not just switch to `k_panic()`? That should resolve in the compiler injecting a controlled fault that can resolve to a PC that can then be used to track down the location in the source code that triggered the `k_panic()`. Could you try that and see if it works and yields better ROM footprint?
> > CC @ioannisg @andrewboie
> 
> @carles: So what we would like to do is to have CONFIG_BT_ASSERT_VERBOSE similar for __ASSERT and NET_BUF then?

Yes, that would be a good thing, then we could standardize on one single way of handling this in all subsystems. I think it makes sense to give users fine-grained control over assertions:

- Assertions disabled
- Assertions enabled, trigger `k_panic()` directly, resolve with `addr2line`
- Assertions enabled, `printk` file and line number
- Assertions enabled, `printk` file, line number and assertion test that failed

carlescufi commented on subsys/bluetooth/common/log.h:
```diff
@@ -53,11 +53,11 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL);
-				BT_ASSERT_PRINT("assert: '" #cond \
-						"' failed\n"); \
+#define BT_ASSERT(cond) do { if (!(cond)) { \
+				BT_ASSERT_PRINT("assert: %s:%d\n", __FILE__, \
```
Why not use `__ASSERT_LOC` here instead of reinventing it?

pfalcon commented on subsys/debug/Kconfig:
```diff
@@ -164,6 +164,13 @@ config ASSERT_NO_FILE_INFO
+config ASSERT_NO_COND_INFO
+	bool "Disable condition info for asserts"
+	help
+	  This option removes condition expression of the assertion that
```
This sounds a bit unclear to me. Can you reword to something along the lines of "Remove condition expression from the printed assertion message...". Or "Don't print condition expression in the assertion message..."

joerchan replied:
Will improve description.

pfalcon commented on subsys/net/buf.c:
```diff
@@ -26,8 +26,8 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #define NET_BUF_INFO(fmt, ...) LOG_INF(fmt, ##__VA_ARGS__)
-#define NET_BUF_ASSERT(cond) do { if (!(cond)) {			  \
-			NET_BUF_ERR("assert: '" #cond "' failed"); \
+#define NET_BUF_ASSERT(cond) do { if (!(cond)) {                             \
```
I would question why/do we still need an adhoc assert just for net_buf. @jukkar? Assuming it's still needed, I'd rather see __ASSERT_COND_INFO() applied here, instead of unconditional change like that. Printing file/line is of course good.

jukkar replied:
> I would question why/do we still need an adhoc assert just for net_buf. @jukkar?

I am ok if we remove these, @jhedberg do you agree?

jhedberg replied:
Yes, I'm fine with removing

joerchan replied:
Removed all NET_BUF_ASSERTs and replaced them with __ASSERT_NO_MSG instead.

joerchan commented:
@jukkar @pfalcon @jhedberg
So removing NET_BUF_ASSERT and using __ASSERT_NO_MSG would mean that we cannot enable asserts for just NET_BUF. I guess you are fine with that?

Should we perhaps do the same for BT_ASSERT? It seems that we use both in BT, which seems just confusing.

@carlescufi Putting a k_panic into every assert is not a huge space-saver it seems. The k_panic "function" would resolve to 4 assembly lines into every assert statement on ARM. So using that would  probably result in the same code-usage as the assert handling we have now.
The assert handler `assert_post_action` will result in a k_panic or a k_oops depending on the context. 

I would suggest we do the following.
Put __LINE__ under CONFIG_ASSERT_NO_FILE_INFO, no point in having line number without the file it belongs to.
Put __ASSERT_LOC under a new define which would remove the printing of ASSERTION FAIL completely.
Put the printk statement in __ASSERT under a define to remove all the messages.

If the user then wishes to have minimal asserts they can disable everything. and the assert will resolve to a simple
`if (test) { assert_post_action("", 0); }`
The user can then implement their own assert_post_action and retrieve the link register to get the assert location.

carlescufi commented:
> @carlescufi Putting a k_panic into every assert is not a huge space-saver it seems. The k_panic "function" would resolve to 4 assembly lines into every assert statement on ARM. So using that would probably result in the same code-usage as the assert handling we have now.

How so? there is not file/line information stored at all in that case, so that should make a difference, given that we are able to retrieve the location without having to store it?

joerchan commented:
> How so? there is not file/line information stored at all in that case, so that should make a difference, given that we are able to retrieve the location without having to store it?

Because making all the assert information configurable, there will be the same amount of code in order to call assert_post_action. It should be possible to use the return register there to see assert location. So we can get the same result. But without losing the k_panic/k_oops policy.

joerchan committed:
kernel: panic: Option ASSERT_NO_FILE_INFO controls panic or oops print

Use option ASSERT_NO_FILE_INFO to control panic or oops location print.
The cause of the exception can be backtraced using the stackframe
instead, which would give the user a way to reduce the footprint of the
panic implementation.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
assert: Move __ASSERT_LOC so it can be used when asserts are disabled

Move __ASSERT_LOC macro so that it can be used by other modules even
when CONFIG_ASSERT are disabled.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
assert: Add verbose option to assertion messages

Add verbose option which would control if the assertion mechanism prints
any information at all. With this disabled they application will have to
use the stack-frame to locate the assertion location.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
assert: Add option to disable assertion messages

Add option to disable the assertion message, this makes all __ASSERT
behave as __ASSERT_NO_MSG instead.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
assert: Disable line number printing together with file printing

Disable printing the line number in assertions when file name has been
disabled. Knowing the line number is not very useful when the name of
the file is unknown.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

aescolar commented on include/sys/__assert.h:
```diff
@@ -20,12 +20,38 @@
+
 #ifdef CONFIG_ASSERT_NO_FILE_INFO
 #define __ASSERT_FILE_INFO ""
+#define __ASSERT_LINE_INFO 0
```
What would be the benefit of removing the line number from the message and replacing it with 0? (Are you hoping the printk string will be optimized into a constant string with less chars?)

joerchan replied:
I wanted to not have all the numbers in of the asserts in the final image. But it seems that this doesn't make any difference as now there are just a whole lot of zeroes instead. I would have to do more to optimize it away it seems. Which would probably include removing the arguments from the assert_post_action function.

joerchan replied:
Completely removing file and line gives me around 3656 bytes saved for the shell. It's roughly the same for the peripheral sample. Around 3k.

joerchan committed:
assert: Completely remove file info and condition expression

Completely remove the file info and condition expression from the
the print statement if they are not enabled. This saves a little code
space which adds up when there are many assert calls.

In bluetooth shell test this saves around 4.5k bytes.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

mped-oticon commented on include/sys/__assert.h:
```diff
@@ -20,12 +20,38 @@
+#define __ASSERT_LINE_INFO 0
 #else  /* CONFIG_ASSERT_NO_FILE_INFO */
 #define __ASSERT_FILE_INFO __FILE__
+#define __ASSERT_LINE_INFO __LINE__
```
It would be cool (or another PR) to make this slightly finer grained: That we can indepdently change if we want FILE or LINE. FILE is overkill, but LINE can be helpful when the compiler is a bit too intelligent.

joerchan replied:
Do you think that is needed? Lines add a lot of code-space, more than file info (I think, haven't checked yet). So when you want file info you might as well have both. How do you intend to only use line?  or only file?

mped-oticon replied:
Correct, but lines will typically be embedded as immediate values and thus occupy PMEM/.text, where FILE will most likely be DMEM/.rodata.
But Oticon may be the only customer for this. No need to fix.

mped-oticon commented on subsys/net/buf.c:
```diff
@@ -78,8 +74,8 @@ static inline struct net_buf *pool_get_uninit(struct net_buf_pool *pool,
-	NET_BUF_ASSERT(buf->flags == 0U);
-	NET_BUF_ASSERT(buf->frags == NULL);
+	__ASSERT_NO_MSG(buf->flags == 0U);
+	__ASSERT_NO_MSG(buf->frags == NULL);
```
I'd vote for mapping `NET_BUF_ASSERT` onto `__ASSERT_NO_MSG`, perhaps that is less politically contentious and still permits netbuf-specifics if so needed.

joerchan replied:
I didn't seem like it was politically contentious. All voted for removing, and I got no reply to my follow up question. So seems removing completely was fine.

mped-oticon replied:
Alright :+1:

joerchan committed:
tests: net: buf: Fix byteorder test writing beyond buffer

Fix byteorder test writing past the data pointer. This would otherwise
have been caught by the assert, but net buf asserts were not enabled in
the test. Reset the buffer between different tests to re-use the same
16-byte buffer.

Turning on asserts gives:

starting test - net_buf_test_byte_order
ASSERTION FAIL [net_buf_simple_tailroom(buf) >= len] @
	ZEPHYR_BASE/subsys/net/buf.c:775
E: r0/a1:  0x00000004  r1/a2:  0x00000307  r2/a3:  0x00000000
E: r3/a4:  0x00006d79 r12/ip:  0x00000000 r14/lr:  0x000028ad
E:  xpsr:  0x61000000
E: Faulting instruction address (r15/pc): 0x000050b0
E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
tests: net: arp: Fix net_buf assert in ARP unit test

Fix ASSERTs that appears when enabling asserts in net buf.
Asserts are:
 - Pulling from net buf before any data has been added.
 - Pulling more data than has been allocated for the buf.

Fix warning:
W: You have 1 IPv4 net_if addresses but 2 network interfaces
W: Consider increasing CONFIG_NET_IF_MAX_IPV4_COUNT value.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
tests: net: ipv6: Fix net_buf asserts in fragment unittest

Fix ASSERTs that appears when enabling asserts in net buf when adding
more data than has been allocated for the buf.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
@jukkar @pfalcon @tbursztyka I removed NET_BUF_ASSERT in favour of __ASSERT_NO_MSG, but that means that now these asserts are enabled when running tests. This is good in my opinion. but it meant that I had to fix a couple of asserts that was being hit in the tests. 
I have fixed those as best as I can. I don't know enough about the ARP protocol to tell if it is correct. But double ETH header looked suspicious to me, but I couldn't make the test pass without it (there is i similar place that pulls from the request instead, but doing that change the test failed).

jukkar commented on tests/net/arp/src/main.c:
```diff
@@ -176,12 +176,14 @@ static inline struct net_pkt *prepare_arp_reply(struct net_if *iface,
 	struct net_eth_hdr *eth;
 
 	pkt = net_pkt_alloc_with_buffer(iface, sizeof(struct net_eth_hdr) +
+					sizeof(struct net_eth_hdr) +
```
Having extra ethernet header here is indeed suspicious. After looking the code, the reason for this is that the test creates the reply packet to the same buffer as where the incoming ARP is in. That is why there is two `net_buf_pull(pkt->buffer, sizeof(struct net_eth_hdr));` calls in that function. Not sure if this is worth changing, the code seems to work ok although is a bit weird looking.

joerchan replied:
I thought it was worth changing, but I wasn't able to make the test pass so I must be missing some detail.

jukkar reviewed (commented):
The networking test changes look ok. The commit subjects in two them are a bit off, could you change them to `tests: net: buf: ....` and `tests: net: arp: ....`, the IPv6 test change commit is named properly,

joerchan commented:
> The networking test changes look ok. The commit subjects in two them are a bit off, could you change them to `tests: net: buf: ....` and `tests: net: arp: ....`, the IPv6 test change commit is named properly,

Thanks. Updated.

carlescufi commented:
@andrewboie since you introduced `k_panic` and `k_oops`, could you take a look at this PR and give feedback?

andrewboie commented on lib/os/assert.c:
```diff
@@ -22,10 +22,16 @@
 __weak void assert_post_action(const char *file, unsigned int line)
+#endif
 {
+#ifndef CONFIG_ASSERT_NO_FILE_INFO
```
Maybe we should just delete the file/line parameters since no configuration currently uses them.

joerchan replied:
Isn't this handler supposed to be overridable by the application so external projects might be using it, I'm hesitent to remove it.

I'm referring to this:
```
 * System designers may wish to substitute this implementation to take other
 * actions, such as logging program counter, line number, debug information
 * to a persistent repository and/or rebooting the system.
```

andrewboie commented on subsys/debug/Kconfig:
```diff
@@ -181,13 +181,37 @@ config FORCE_NO_ASSERT
+	help
+	  This option removes the additional message from the printed assert.
+	  Enabling this will save target cod space, and should be disabled
+	  before disabling file and line information as the message can be
```
> should be disabled before disabling file and line information

Can we express this dependency in Kconfig itself?

joerchan replied:
Not sure how. But I meant this as a recommendation, and not a rule. I can update the text with "is recommended to" instead of "should"

andrewboie commented on include/kernel.h:
```diff
@@ -4920,7 +4920,7 @@ extern void z_sys_power_save_idle_exit(s32_t ticks);
  */
 #define z_except_reason(reason) do { \
-		printk("@ %s:%d:\n", __FILE__,  __LINE__); \
+		__ASSERT_LOC("Exception");\
```
What this is going to print is "ASSERTION FAIL [Exception] @ file:line"
But z_except_reason() is called by k_panic/k_oops which are *not* assertions.
I'd rather the output did not include the word "assertion" in it, because it is not.

tyrantlizard77 reviewed (changes requested):
Broadly speaking this looks OK (I have no comment on the BT/Net code) but the printed output on k_oops()/k_panic() isn't right

jukkar commented on subsys/debug/Kconfig:
```diff
@@ -181,13 +181,37 @@ config FORCE_NO_ASSERT
+	bool "Disable message for asserts"
+	help
+	  This option removes the additional message from the printed assert.
+	  Enabling this will save target cod space, and should be disabled
```
Typo s/cod/code/

joerchan replied:
Fixed

joerchan commented:
@andrewboie Thank you for the feedback. I have updated k_panic/k_oops to print the same as before, but is controlled by the CONFIG_ASSERT_NO_FILE_INFO and CONFIG_ASSERT_VERBOSE options in the case where the application might want to trim the code size of the executable, and instead backtrace the exception through other methods, or ignore it.
